beacon-chain/cache/sync_committee.go

```diff
@@ -0,0 +1,163 @@
+type syncCommitteeIndexPosition struct {
+	currentSyncCommitteeRoot [32]byte
+	vIndexToPositionMap      map[[48]byte]*positionInCommittee
+}
+
+// Index position of individual validator of current epoch and previous epoch sync committee.
+type positionInCommittee struct {
+	currentEpoch []uint64
+	nextEpoch    []uint64
+}
+
+// NewSyncCommittee initializes and returns a new SyncCommitteeCache.
+func NewSyncCommittee() *SyncCommitteeCache {
+	return &SyncCommitteeCache{
+		cache: cache.NewFIFO(keyFn),
+	}
+}
+
+// CurrentEpochIndexPosition returns current epoch index position of validator pubKey with respect with sync committee.
+// If the input pubKey has no assignment, an empty list will be returned.
+// If the input root does not exist in cache, ErrNonExistingSyncCommitteeKey is returned. Then performing manual checking of state for index position in state is recommended.
+func (s *SyncCommitteeCache) CurrentEpochIndexPosition(root [32]byte, pubKey [48]byte) ([]uint64, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	pos, err := s.idxPositionInCommittee(root, pubKey)
+	if err != nil {
+		return nil, err
+	}
+	if pos == nil {
+		return []uint64{}, nil
+	}
+
+	return pos.currentEpoch, nil
+}
+
+// NextEpochIndexPosition returns next epoch index position of validator pubKey in respect with sync committee.
+// If the input pubKey has no assignment, an empty list will be returned.
+// If the input root does not exist in cache, ErrNonExistingSyncCommitteeKey is returned. Then performing manual checking of state for index position in state is recommended.
+func (s *SyncCommitteeCache) NextEpochIndexPosition(root [32]byte, pubKey [48]byte) ([]uint64, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	pos, err := s.idxPositionInCommittee(root, pubKey)
+	if err != nil {
+		return nil, err
+	}
+	if pos == nil {
+		return []uint64{}, nil
+	}
+	return pos.nextEpoch, nil
+}
+
+// Helper function for `CurrentEpochIndexPosition` and `NextEpochIndexPosition` to return a mapping
+// of validator pubKey to its index(s) position.
+func (s *SyncCommitteeCache) idxPositionInCommittee(root [32]byte, pubKey [48]byte) (*positionInCommittee, error) {
+	obj, exists, err := s.cache.GetByKey(key(root))
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, ErrNonExistingSyncCommitteeKey
+	}
+	item, ok := obj.(*syncCommitteeIndexPosition)
+	if !ok {
+		return nil, errNotSyncCommitteeIndexPosition
+	}
+	idxInCommittee, ok := item.vIndexToPositionMap[pubKey]
+	if !ok {
+		return nil, nil
+	}
+	return idxInCommittee, nil
+}
+
+// UpdatePositionsInCommittee updates caching of validators position in sync committee in respect to
+// current epoch and next epoch. This should be called when `current_sync_committee` and `next_sync_committee`
+// change and that happens every `EPOCHS_PER_SYNC_COMMITTEE_PERIOD`.
+func (s *SyncCommitteeCache) UpdatePositionsInCommittee(state iface.BeaconStateAltair) error {
+	s.lock.Lock()
```

Comment:
In case this method becomes really expensive, you can lock just around the for loops as an optimization

Reply:
Addressed. Thanks!